conn.go

```diff
@@ -283,40 +314,61 @@ func (l *Conn) processMessages() {
 		select {
 		case l.chanMessageID <- messageID:
 			messageID++
-		case messagePacket, ok := <-l.chanMessage:
+		case message, ok := <-l.chanMessage:
 			if !ok {
 				l.Debug.Printf("Shutting down - message channel is closed")
 				return
 			}
-			switch messagePacket.Op {
+			switch message.Op {
 			case MessageQuit:
 				l.Debug.Printf("Shutting down - quit message received")
 				return
 			case MessageRequest:
 				// Add to message list and write to network
-				l.Debug.Printf("Sending message %d", messagePacket.MessageID)
-				l.chanResults[messagePacket.MessageID] = messagePacket.Channel
-				// go routine
-				buf := messagePacket.Packet.Bytes()
+				l.Debug.Printf("Sending message %d", message.MessageID)
+				l.chanResults[message.MessageID] = message.Channel
+
+				if l.RequestTimeout != 0 {
+					go func() {
+						time.Sleep(l.RequestTimeout)
+						timeoutMessage := &messagePacket{
+							Op:        MessageTimeout,
+							MessageID: message.MessageID,
+						}
+						l.sendProcessMessage(timeoutMessage)
+					}()
+				}
+
+				buf := message.Packet.Bytes()
 
 				_, err := l.conn.Write(buf)
```

Comment:
I would prefer the timeout starting _after_ the request is sent, but I could be persuaded either way.  

My main opinion is that you and @flowhamster have a good handle on getting this feature ready to pull in.  

I'll chime in if I have any random thoughts, but otherwise will enjoy watching you finish it up :)